c/dec/huffman.h

```diff
@@ -33,11 +33,60 @@ static const uint16_t kMaxHuffmanTableSize[] = {
 
 #define BROTLI_HUFFMAN_MAX_CODE_LENGTH_CODE_LENGTH 5
 
+#if !defined(BROTLI_HUFFMAN_CODE_FAST_LOAD)
 typedef struct {
   uint8_t bits;    /* number of bits used for this symbol */
   uint16_t value;  /* symbol value or table offset */
 } HuffmanCode;
 
+/* To allow optimization of HuffmanCode accesses in hot paths,
+ * please use the following macros to:
+ *
+ *  - prepare a HuffmanCode pointer for access
+ *     ( BROTLI_HC_PREPARE_TO_ACCESS(HuffmanCode*) )
+ *  - update the pointer each time it is changed
+ *     ( BROTLI_HC_UPDATE(HuffmanCode*) )
+ *  - get the bits or value fields
+ *     ( BROTLI_HC_GET_BITS/VALUE(HuffmanCode*) )
+ *
+ * For example, assuming |table| contains a HuffmanCode pointer:
+ *
+ *   BROTLI_HC_PREPARE_TO_ACCESS(table);
+ *   BROTLI_HC_UPDATE(table);
+ *   *bits = BROTLI_HC_GET_BITS(table);
+ *   *value = BROTLI_HC_GET_VALUE(table);
+ *   table += offset;
+ *   BROTLI_HC_UPDATE(table);
+ *   *bits2 = BROTLI_HC_GET_BITS(table);
+ *   *value2 = BROTLI_HC_GET_VALUE(table);
+ *
+ */
+
+#define BROTLI_HC_PREPARE_TO_ACCESS(H)
+#define BROTLI_HC_UPDATE(H)
+
+/* Getters must be given a pointer to a HuffmanCode. */
+#define BROTLI_HC_GET_BITS(H) (H->bits)
+#define BROTLI_HC_GET_VALUE(H) (H->value)
+
+/* Setters must be given a HuffmanCode. */
+#define BROTLI_HC_SET_BITS(H, B) H.bits = B
+#define BROTLI_HC_SET_VALUE(H, V) H.value = V
+#else
+typedef BROTLI_ALIGNED(4) uint32_t HuffmanCode;
+
+#define BROTLI_HC_PREPARE_TO_ACCESS(H) uint32_t __fastload_##H
```

Comment:
Going to check if we could get rid of declaration/fetch wrappers.
If in "non-fast" mode copying struct to local variable is not slower (hopefully, so), then "prepare" would be plain HuffmanCode declaration, and "update" would be assignment...

Reply:
Hi @eustas, thanks for the feedback.

So just to check, you're suggesting we make both the setters and getters access HuffmanCode (not a pointer):

```
/* Getters and setters must be given a HuffmanCode. */        
#define BROTLI_HC_GET_BITS(H) (H.bits)                        
#define BROTLI_HC_GET_VALUE(H) (H.value)
```

(and with FAST_LOAD:)

```
/* Getters and setters must be given a HuffmanCode. */   
#define BROTLI_HC_GET_BITS(H) (H & 0xFFu)                
#define BROTLI_HC_GET_VALUE(H) (H >> 16)
```

And then we drop the PREPARE_TO_ACCESS and UPDATE macros, and just do things like:

```
static BROTLI_INLINE uint32_t DecodeSymbol(uint32_t bits,                  
                                           const HuffmanCode* table,       
                                           BrotliBitReader* br) {          
  HuffmanCode huffman_code;                                                
  table += bits & HUFFMAN_TABLE_MASK;                                      
  huffman_code = *table;                                                   
  if (BROTLI_HC_GET_BITS(huffman_code) > HUFFMAN_TABLE_BITS) {             
    uint32_t nbits = BROTLI_HC_GET_BITS(huffman_code) - HUFFMAN_TABLE_BITS;
    BrotliDropBits(br, HUFFMAN_TABLE_BITS);                                
    table += BROTLI_HC_GET_VALUE(huffman_code);                            
    table += (bits >> HUFFMAN_TABLE_BITS) & BitMask(nbits);                
    huffman_code = *table;                                                 
  }                                                                        
  BrotliDropBits(br, BROTLI_HC_GET_BITS(huffman_code));                    
  return BROTLI_HC_GET_VALUE(huffman_code);                                
}
```

(Which name would you prefer for the local variable? `huffman_code`, `prefix_code`, `_table`, etc?

Reply:
@eustas my rewrite of the code as described above unfortunately leads to code changes on x86, and a minor regression on the x86 chromebook I'm testing on, so I'm not sure if we can get rid of the prepare/update wrappers.